### ISSUE TITLE
toctree contains reference to nonexisting document 

### ISSUE THREAD

**Issue by amueller** [numpy/numpydoc#69]
I'm still trying to migrate scikit-learn to numpydoc master.
I'm having some trouble with this issue:
http://stackoverflow.com/questions/12206334/sphinx-autosummary-toctree-contains-reference-to-nonexisting-document-warnings

Apparently something goes wrong with when autosummary is called?
That didn't use to be the case.
I can't really do `numpydoc_show_class_members = False` because then there is no method table any more :(

**Comment by amueller** [comment]
so setting ``class_members_toctree = False`` solves my problem. But there still seems to be a bug...

**Comment by jorisvandenbossche** [comment]
For pandas we used the same 'hack' as in the numpy docs: https://raw.githubusercontent.com/numpy/numpy/master/doc/source/_templates/autosummary/class.rst, but I am not sure anymore if this was to solve those error messages (long time ago I did this for the pandas docs)

**Comment by amueller** [comment]
huh, we didn't need this hack, the ``numpydoc_show_class_members = True`` combine with ``class_members_toctree = False`` seems to produce the right results. It just seems kinda odd if there is only one specific configuration that is needed to make it actually work... well better than a hack.

**Comment by stefanv** [comment]
That is weird; if you figure out what's going on, let's put a note in
the README at least.

**Comment by amueller** [comment]
I'm kinda punting on it for now. This configuration Works For Me (TM).

**Comment by jnothman** [comment]
So is this suggesting that the `.. autosummary` in numpydoc-mangled docstrings simply isn't working for autogeneration, hence the need for the [above hacks](https://raw.githubusercontent.com/numpy/numpy/master/doc/source/_templates/autosummary/class.rst)?

This is very believable to me: I can't see anywhere in autosummary where it sees the mangled docs. It's either processing through [the `autosummary_generate` option](https://github.com/sphinx-doc/sphinx/blob/5ae72bdd28075a02213ede26accd196c1a9dba05/sphinx/ext/autosummary/__init__.py#L591), at a builder-inited hook, which operates over paths,  or `autogen` is run manually [again on paths](https://github.com/sphinx-doc/sphinx/blob/5ae72bdd28075a02213ede26accd196c1a9dba05/sphinx/ext/autosummary/generate.py#L371).

Now we no longer use `.. autosummary` for display of attributes, only methods, so we should rid of the hack use of autosummary from attribute lists, then. I'm not sure if `class_members_toctree ` is having useful effect for method listing? (Or is it possible someone's been using numpydoc directly as a pre-process rather than as an `autodoc-process-docstring` hook?)

**Comment by jnothman** [comment]
So I think we basically need to recommend a hack at the template level, and that `class_members_toctree` should then match what the hack does :\

**Comment by mattip** [comment]
I am willing to work on this, but as I am new to the codebase would need some help parsing @jnothman 's [comment](https://github.com/numpy/numpydoc/issues/69#issuecomment-338878979) above. Is there a general design document that describes what autosummary does and when? It seems there are three options for generating summaries:

- the `autosummary_generate` option (I guess this is what numpy does since we have `autosummary_generate = glob.glob("reference/*.rst")` in `conf.py`)
- a builder-initiated hook which operates over paths
- `autogen` is run manually again on paths

What do the second and third one mean? Should one use be preferred over the others?

Then there is the sentence below, what is meant by "the hack use" and what should replace it? 

> Now we no longer use `.. autosummary` for display of attributes, only methods, so we should rid of the hack use of autosummary from attribute lists

**Comment by mattip** [comment]
FWIW, adding these lines to the `conf.py` does not silence the warnings in numpy
```
numpydoc_show_class_members = True
class_members_toctree = False
```

**Comment by jnothman** [comment]
I'm not really sure what to say here, and my recall is weak. The different ways to run autosummary are not really relevant as long as they only collect names to autodoc from reST files, rather than from autodoc-generated reST. I suppose you could consider fixing that in autodoc. Otherwise I'm not sure how to go about fixing this.

**Comment by mattip** [comment]
Removing the nested directive from the template for class methods results in the method having two links in the output document where the attribute has only one, so something is different in the handling of class methods and attributes.

**Comment by mattip** [comment]
maybe relate to #74 ?

**Comment by mattip** [comment]
Nope, my "fix" is garbage. The attributes show up in the currently rendered documents, but formatted badly and above the methods. It seems the nested directive hack is needed to prevent duplicate rendering of both the Methods and the Attributes and that something else is happening.

**Comment by jnothman** [comment]
Formatted "badly" is a matter of taste. They're formatted just like Parameters and Returns. I.e. they're formatted as numpydoc formats things, rather than according to the decisions of autosummary.

**Comment by mattip** [comment]
Yes, sorry, I let my frustration through. My fault for not noticing the `Attributes` do appear in the rendered document, but formatted differently than the `Methods`.

**Comment by jorisvandenbossche** [comment]
It's related to the longish discussion I had here https://github.com/numpy/numpydoc/pull/106#issuecomment-343220139. 
I still think it would be good if numpydoc would want to support this usecase of keeping the Attributes listing as they were before for downstream users as pandas (and numpy? don't know if they want to keep it)

I have a PR that does that (https://github.com/numpy/numpydoc/pull/161), but of course, that's still a bit independent from the discussion in this issue about the hack in the class template and the warnings (for this, in pandas we long time ago just copied the hack from numpy). I would also need to look at it again to again understand why it was needed.

**Comment by mattip** [comment]
Maybe I have made progress. I created templates for methods, members,and attributes under `autosummary`, copying the `base.rst` but adding a top line `:orphan:`. While not impacting the output (as far as I can see) it does quiet the warning. So here is what I have so far:

- The hack in the `class.rst` template is needed to prevent the `Methods` and `Attributes` sections from appearing twice in the class rendered document. 
- The hack also causes the members and attributes label to turn into links (and generates corresponding pages). See for instance the page for [numpy.broadcast](http://www.numpy.org/devdocs/reference/generated/numpy.broadcast.html)
- Class members that are neither attributes nor methods are listed with the attributes, but are not rendered as links.
- A class method [page](http://www.numpy.org/devdocs/reference/generated/numpy.broadcast.reset.html) is rendered with a sidebar and breadcrumbs across the top
- A class [atrribute page](http://www.numpy.org/devdocs/reference/generated/numpy.broadcast.index.html) is rendered with a truncated breadcrumb and no sidebar 
- A class member [page](http://www.numpy.org/devdocs/reference/generated/numpy.broadcast.nd.html) is created, even though it has no link from the parent class page. It appears in the search results.

I still do not understand why  the `class.rst` template hack generates warning NumPy but not for SciPy . Both use exactly the same template and have very similar `conf.py` files, although with slightly different extension modules.

**Comment by mattip** [comment]
Could someone with commit permissions reopen since the core issue (hack needed in class template, which then causes many warnings) is not yet fixed

**Comment by rgommers** [comment]
wasn't the point of that PR to remove the warnings? not sure I understand your comment correctly, but if it's about tracking `orphan` so it can be removed again later, I think that's better as a new issue

**Comment by mattip** [comment]
Ok, sorry for the noise. Close this and I will open a new one.

**Comment by rgommers** [comment]
Ah no, my apologies - I thought we were in https://github.com/numpy/numpy/issues/11316. NumPy has a workaround now, but `numpydoc` still needs a proper fix.

**Comment by ReenigneArcher** [comment]
Just came across these warnings. Not sure if there's a better way, but I solved it by manually adding the `Methods` heading to my class docstring. Kind of redundant, but at least docs build without the warnings now. 

https://stackoverflow.com/a/77588774/11214013

Docs actually built fine before, except the warnings generated... but readthedocs builds are set to fail on any warnings which I think is a good practice to keep in place.
